text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,802 @@
+
+fn send_file(stream: &UnixStream, file: std::fs::File) -> std::io::Result<()> {
+    let mut ancillary = AncillaryData::new();
+    ancillary.add_file_descriptors(&[
+        file.as_fd(),
+    ]);
+
+    MessageSender::new(stream, b"\x00")
+        .ancillary_data(&mut ancillary)
+        .send(&stream)?;
+    Ok(())
+}
+```
+
+Receiving a file descriptor:
+
+```rust
+use std::os::unix::net::{AncillaryData, MessageReceiver, UnixStream};
+
+fn recv_file(stream: &UnixStream) -> std::io::Result<std::fs::File> {
+    // TODO: expose CMSG_SPACE() in a user-friendly way.
+    const ANCILLARY_CAPACITY: usize = 100;
+
+    let mut ancillary = AncillaryData::with_capacity(ANCILLARY_CAPACITY);
+    let mut buf = [0u8; 1];
+    MessageReceiver::new(stream, &mut buf)
+        .ancillary_data(&mut ancillary)
+        .recv(&stream)?;
+
+    // TODO: error handling (std::io::Error if not enough FDs returned)
+    let mut owned_fds = ancillary.take_owned_fds().unwrap();
+    let sent_fd = owned_fds.swap_remove(0);
+    Ok(sent_fd.into())
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Control messages
+
+Each control message is represented as a `ControlMessage` struct, which holds
+a reference to the underlying data.
+
+```rust
+struct ControlMessage<'a>;
+
+impl ControlMessage<'a> {
+    fn new(
+        cmsg_level: c_int,
+        cmsg_type: c_int,
+        data: &'a [u8],
+    ) -> ControlMessage<'a>;
+
+    // Opaque platform-specific integers; same as `struct cmsghdr` fields.
+    fn cmsg_level(&self) -> c_int;
+    fn cmsg_type(&self) -> c_int;
+
+    // The type-specific data of this control message.
+    fn data(&self) -> &[u8];
+
+    // Whether this control message is truncated, such as by being received
+    // into a too-short buffer.
+    fn truncated(&self) -> bool;
+}
+```
+
+A `&ControlMessages` represents a buffer full of control messages. It can be
+inspected and iterated over to obtain `ControlMessage` values.
+
+
+```rust
+struct ControlMessages;
+
+impl ControlMessages {
+    fn as_bytes(&self) -> &[u8];
+    fn is_empty(&self) -> bool;
+    fn len(&self) -> usize;
+}
+
+impl<'a> IntoIterator for &'a ControlMessages {
+    type Item = ControlMessage<'a>;
+    type IntoIter = ControlMessagesIter<'a>;
+}
+
+struct ControlMessagesIter<'a> { ... }
+
+impl<'a> Iterator for ControlMessagesIter<'a> {
+    type Item = ControlMessage<'a>;
+}
+```
+
+A `ControlMessagesBuf` is the owned variant of `ControlMessages`. It exposes a
+subset of the `Vec` capacity management functions, with a public API that only
+allows operations that won't reduce its length (since that would risk losing
+information about received file descriptors).
+
+```rust
+struct ControlMessagesBuf;
+
+impl ControlMessagesBuf {
+    fn new() -> ControlMessagesBuf;
+    fn with_capacity(capacity: usize) -> ControlMessagesBuf;
+
+    fn capacity(&self) -> usize;
+
+    fn push(&mut self, message: &ControlMessage<'_>);
```

Comment:
> This probably needs a `try_push` or `push_with_capacity` if the sender wants to ensure that the data stays within a certain size so that the receiver doesn't have to resize theirs.

A `try_push` would probably be better handled by `try_reserve` + `push`, as with `Vec`.

The `ControlMessagesBuf` doesn't limit how large the buffer is allowed to grow, since the number and size of control messages is platform-dependent. Code that needs to enforce a limit on some resource (for example to send only a single FD at a time) would need to do that separately.

Note that ancillary data may be truncated, so a receiver doesn't need to be written to accept arbitrary amounts of it. If a receiver expects to receive a single FD, it only needs to allocate enough space to hold that.

> Also, this could use some words on the lifetimes. Is sending supposed to work with borrowed FDs? If so the whole struct needs a lifetime so that it's guaranteed that the FDs outlive the struct's lifetime. But that makes buffer recycling more complicated

See the `AncillaryData` struct, which has a lifetime for borrowed FDs. FD ownership isn't part of the `ControlMessages` buffer management.